File: modules/container/test/test_consecutive_pair_container.py
import IMP
import IMP.test
import IMP.container
import IMP.display
import math

class Tests(IMP.test.TestCase):
    """Tests for tunnel scores"""
    def test_connectivity_zero_set(self):
        """Test consecutive pair container decomposition"""
        m= IMP.Model()
        ps=[IMP.Particle(m) for i in range(0,15)]
        ds=[IMP.core.XYZR.setup_particle(p) for p in ps]
        bb= IMP.algebra.BoundingBox3D(IMP.algebra.Vector3D(0,0,0),
                                      IMP.algebra.Vector3D(10,10,10))
        for d in ds:
            d.set_coordinates(IMP.algebra.get_random_vector_in(bb))
            print d
            d.set_radius(4)
        cpc= IMP.container.ConsecutivePairContainer(ps)
        hdps= IMP.core.HarmonicDistancePairScore(0,1)
        r= IMP.container.PairsRestraint(hdps, cpc)
        m.add_restraint(r)
        self.assert_(r.evaluate(False) > 0)
        rd= r.create_decomposition()
        rds= IMP.get_restraints([rd])
        self.assertEqual(len(rds), len(ps)-1)
        re= r.evaluate(False)
        rde= rd.evaluate(False)
        rdse= [x.evaluate(False) for x in rds]
        self.assertAlmostEqual(re, rde, delta=.1)
        self.assertAlmostEqual(re, sum(rdse), delta=.1)
        print re, rde, sum(rdse)

    def test_pair_filter(self):
        """Test consecutive pair filter"""
        m= IMP.Model()
        ps=[IMP.Particle(m) for i in range(0,15)]
        ecpc= IMP.container.ExclusiveConsecutivePairContainer(ps)
        ef= IMP.container.ExclusiveConsecutivePairFilter();
        for i in range(1,len(ps)):
            self.assertEquals(ef.get_value((ps[i-1], ps[i])), 1)
        for i in range(2,len(ps)):
            self.assertEquals(ef.get_value((ps[i-2], ps[i])), 0)




if __name__ == '__main__':
    IMP.test.main()

File: modules/container/test/test_consecutive_pair_container.cpp
/**
 *  \file test_.cpp   \brief A class to represent a voxel grid.
 *
 *  Copyright 2007-2021 IMP Inventors. All rights reserved.
 *
 */
#include <IMP/container/ConsecutivePairContainer.h>
#include <IMP/algebra/BoundingBoxD.h>
#include <IMP/core/DistancePairScore.h>
#include <IMP/algebra/vector_generators.h>
#include <IMP/RestraintSet.h>
#include <IMP/core/XYZ.h>
#include <IMP/flags.h>
#include <algorithm>
#include <IMP/container/generic.h>

int main(int argc, char *argv[]) {
  IMP::setup_from_argv(argc, argv, "Test ConsecutivePairContainer.");

  IMP_NEW(IMP::Model, m, ());
  IMP::ParticleIndexes ps;
  IMP::algebra::BoundingBox3D bb(IMP::algebra::Vector3D(0, 0, 0),
                                 IMP::algebra::Vector3D(10, 10, 10));
  for (unsigned int i = 0; i < 15; ++i) {
    ps.push_back(m->add_particle("P"));
    IMP::core::XYZ::setup_particle(m, ps.back(),
                                   IMP::algebra::get_random_vector_in(bb));
  }
  IMP_NEW(IMP::container::ConsecutivePairContainer, cpc, (m, ps));
  IMP_NEW(IMP::core::HarmonicDistancePairScore, hdps, (0, 1));
  IMP::Pointer<IMP::Restraint> r =
      IMP::container::create_restraint(hdps.get(), cpc.get());
  IMP_USAGE_CHECK(r->evaluate(false) > 0, "zero evaluate");
  IMP::Pointer<IMP::Restraint> rd = r->create_decomposition();
  IMP::RestraintsTemp rds = IMP::get_restraints(IMP::RestraintsTemp(1, rd));
  IMP_USAGE_CHECK(rds.size() == ps.size() - 1,
                  "Bad lengths: " << rds.size() << " vs " << ps.size() - 1);
  double re = r->evaluate(false);
  double rde = rd->evaluate(false);
  IMP_CHECK_VARIABLE(re);
  IMP_CHECK_VARIABLE(rde);
  IMP_USAGE_CHECK(std::abs(rde - re) < .1,
                  "Invalid decomposed score: " << re << " vs " << rde);
  return 0;
}
